### PR TITLE
Functions to read/write binary representations of numbers

### PR DESCRIPTION
**Pull request by alainfrisch**
[MPR#5771](https://caml.inria.fr/mantis/view.php?id=5771) introduced primitives to read/write binary representations of 16, 32 and 64-bit integers directly to/from (big)strings (and now bytes), which is very useful for any program that needs to serialize numbers efficiently.  Those primitives are exposed only in the [ocplib-endian](https://github.com/OCamlPro/ocplib-endian) library, but I believe most people would expect to find the functionality in the stdlib, and this is easy since the hard work of supporting the primitives in the compiler and runtime system is behind us.  Moreover, extending Buffer directly allows for more efficient functions than one could get with an external library.

This PR is an attempt to launch the discussion around the addition of such features.  It is far from ready to be merged!  Currently, it adds:

 - Functions to read/write 16-, 32- and 64-bit "binary" integer to `bytes` (only versions with bound checks are exposed).
 - Functions to append 16-, 32- and 64-bit "binary" integer to `Buffer.t`; also single-precision and double-precision floating point numbers.

These function all use the native endianness of the current machine.  This might be problematic for some use cases, but perfectly fine for others (and better, performance-wise), and one could consider adding versions with explicit endianness (as ocp-endian does).

Missing:
  - Functions to read/write floats to/from `bytes` values.
  - Functions to read numbers from `string` values.
  - Functions to read/write floats to/from bigarrays of chars.
  - Variants without bound checks.
  - Perhaps also functions to read/write from/to channels.
 
The performance benefits of the new functions can be illustrated with the following micro-benchmark:
````ocaml
let () =
  let a = Float.Array.create 1000 in
  for j = 0 to Float.Array.length a - 1 do
    Float.Array.set a j (float j)
  done;
  let b = Buffer.create 100 in
  for i = 1 to 100000 do
    Buffer.clear b;
    for j = 0 to Float.Array.length a - 1 do
      Buffer.add_float64_bin b (Float.Array.get a j)
    done;
  done
````

and its variants using:

````ocaml
let add_float64_bin b x =
  let i = Int64.bits_of_float x in
  Buffer.add_char b (Char.unsafe_chr (Int64.to_int (Int64.logand i 0xffL)));
  Buffer.add_char b (Char.unsafe_chr (Int64.to_int (Int64.logand (Int64.shift_right_logical i (8*1)) 0xffL)));
  Buffer.add_char b (Char.unsafe_chr (Int64.to_int (Int64.logand (Int64.shift_right_logical i (8*2)) 0xffL)));
  Buffer.add_char b (Char.unsafe_chr (Int64.to_int (Int64.logand (Int64.shift_right_logical i (8*3)) 0xffL)));
  Buffer.add_char b (Char.unsafe_chr (Int64.to_int (Int64.logand (Int64.shift_right_logical i (8*4)) 0xffL)));
  Buffer.add_char b (Char.unsafe_chr (Int64.to_int (Int64.logand (Int64.shift_right_logical i (8*5)) 0xffL)));
  Buffer.add_char b (Char.unsafe_chr (Int64.to_int (Int64.logand (Int64.shift_right_logical i (8*6)) 0xffL)));
  Buffer.add_char b (Char.unsafe_chr (Int64.to_int (Int64.logand (Int64.shift_right_logical i (8*7)) 0xffL)))
````

The version using the new function runs about 6.5 times faster on my machine.

### CODE REVIEW

**Comment by nojb** [comment]
I strongly support adding these functions to the stdlib.

- I think adding little endian and big endian variants would be quite useful : a very common use-case for these functions is interfacing with binary protocols and/or formats in which case one always needs to use a specific endianness. A suffix `_le` and `_be` could be used. Adding the three versions (little endian, big endian, and host endian) may be optimal but a bit excessive.
- For the case of bigstrings (1-dimensional bigarrays), I think that adding a dedicated `Bigstring` module makes sense, containing roughly the union of operations on `bytes` and `Bigarray`. See also [MPR#7772](https://caml.inria.fr/mantis/view.php?id=7772).
- For naming, is the suffix `_bin` really necessary? What about introducing a signature
```
module type S = sig
  type t
  val get_int16: t -> int
  val get_int32: t -> int32
  val get_int64: t -> int64
  val set_int16: t -> int -> unit
  val set_int32: t -> int32 -> unit
  val set_int64: t -> int64 -> unit
  (* etc *)
end
```
and define modules `Bytes.BE : S with t := bytes`, `Bytes.LE : S with t := bytes`, same for `Buffer` (with `t := Buffer.t`, etc), and so on. Unfortunately, this signature does not work for the case of channels (where the type `t` is different for `get` and `set` functions (maybe splitting the signature in an "input" part and an "output" part would be better to also handle the case of `string`).

**Comment by alainfrisch** [comment]
- I tend to agree that exposing versions with explicit endianness would be good, but I think it's good to keep the version with the native endianness of the machine (which is useful e.g. for inter-process communication on the same machine, or to interact with buffers containing native "C" data).  The number of functions is a bit worrying, though, between int16/int32/int64/float32/float64, le/be/native, string/bytes/bigstring/buffer(/channel), possibly safe/unsafe.

- An alternative could be to expose fewer functions, taking arguments, and hope/check that the inliner will be good enough to eliminate the overhead.  E.g.:
````ocaml
type _ number_ty = Int16: int number_ty | Int32: int32 number_ty | ...
val Buffer.add_number: ?endian:[`LE|`BE](*default:Native*) -> Buffer.t -> 'a number_ty -> 'a -> unit  
````

- About the _bin suffix: I had in mind Stdlib.print_int and co, working with the ASCII decimal representation.  But perhaps we don't care.

- I forgot about input_binary_int/output_binary_int, which we already have: big-endian, 4 bytes, but working with `int`, not `int32`.

**Comment by alainfrisch** [comment]
Yes another approach would be to expose only functions using the native endianness, plus:
````ocaml
 val Int32.Endian.little: int32 -> int32
   (** Modify the value, if needed, so that functions such as Bytes.set_int32 produce a little-endian
        representation of the integer, independently of the native encoding of the host machine. Similarly,
        this function can be applied to the result of functions such as Bytes.get_int32 when reading
        a 32-bit integer in little-endian. *)

 val Int32.Endian.big: int32 -> int32
    (** .... *)
````
(and same for 16-bit and 64-bit)

**Comment by xclerc** [comment]
This looks like a very useful addition.

Even if unsafe versions are not exposed, I think it would make
sense to avoid some bound checks in `Buffer.add_xyz` (they
are avoided in `Buffer.add_char` by using `Bytes.unsafe_set`
rather than `Bytes.set`).

(I would also consider adding `add_int8`/`get_int8`/`set_int8`.
They would of course be mere aliases to existing functions,
but would allow one to write more uniform code when working
on file formats or network protocols.)

**Comment by dbuenzli** [comment]
In my experience what you need most of the time is the explicit endianness (network, codecs, etc.) and it's less error prone if the native endianness has as an explicit suffix; otherwise you tend to use it without thinking about the issue (or it will be the first one to auto-complete...) and you likely end up with a bug.

@alainfrisch I think your last suggestion would make things more confusing than enlighting. Personally I don't mind the number of functions. I think this should be simply:
```
{add,get,set}_int16_{ne,be,le} 
{add,get,set}_int32_{ne,be,le}
{add,get,set}_int64_{ne,be,le}
```
I don't think it's worth adding unsafe versions, people willing to go into unsafe territory can play the primitive game. I don't think it's worth adding the float versions this is an `{Int64,Int32}.{bits_of_float,float_of_bits}` away.

**Comment by alainfrisch** [comment]
>  I don't think it's worth adding the float versions this is an {Int64,Int32}.{bits_of_float,float_of_bits} away.

I was concerned that this would cause extra boxing of the intermediate int64 if the function (e.g. Buffer.add_int64) is not inlined.  To be checked.

**Comment by alainfrisch** [comment]
Ok, it's seem fine to let the user do that float->int64 mapping (checked only with the "Closure" inliner, but I'm sure flambda will be ok), with default inlining settings.

**Comment by alainfrisch** [comment]
I also check that the check for Sys.big_endian is eliminated in e.g. `Buffer.add_int64_le`, and that the performance is the same as `Buffer.add_int64_ne` (on a little-endian system).

For the naming: are _be, _le, _ne good enough?

Concerning the scope of this PR:
 - I suggest to keep the idea of providing channel operations for later.
 - Bigstring, which would be a whole new module, would deserve its own PR.
 - Unsafe versions are indeed not needed (the impact on performance is rather small anyway).
 - Perhaps adding get_int* functions to String makes sense, to keep it in sync with Bytes, even though most people would probably use bytes anyway for parsing binary data, I guess.

**Comment by alainfrisch** [comment]
One more point: it could be useful to provide both get_int16 (sign extension) and get_uint16 (always returning positive integers).

**Comment by dbuenzli** [comment]
Agreed on all points. I think it would be good to add the functions in `String` though.

> One more point: it could be useful to provide both get_int16 (sign extension) and get_uint16 (always returning positive integers).

Yes ! FWIW I have this in an unpublished code of similar functionality. It's the kind of bit fiddling things you prefer to be handled by the library.

File: stdlib/bytes.mli [hunk 1]
```diff
@@ -463,6 +463,91 @@ val of_seq : char Seq.t -> t
 (** Create a string from the generator
     @since 4.07 *)
 
+(** {6 Binary storage of numbers} *)
+
+external get_int16_ne : bytes -> int -> int = "%caml_bytes_get16"
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: IIRC there's no advantage to expose this as an external at the `.mli` level.

Comment (nojb) [reply stdlib/bytes.mli]: Exposing the external in the `.mli` (and hence in the `.cmi`) allows every use of this function to compile a call to the primitive directly (no passing through a wrapper function). Same as inlining, but this works even if inlining does not take place (e.g. in bytecode).

Comment (gasche) [reply stdlib/bytes.mli]: Is this an important property, though? Because the cost is that if we ever decide to change the implementation (for example wrapping it into a bound-checking or exception-changing routine) or primitive name, we break the interface, and that seems quite high to me.

Comment (nojb) [reply stdlib/bytes.mli]: I am not sure I understand: is there a way to write code that would break if one changes the signature from `external` to `val` ?

Comment (damiendoligez) [reply stdlib/bytes.mli]: Some people like to copy the interfaces from the stdlib for their own implementations, and they would get bitten by this.

Other people avoid this problem by `include`ing the interfaces, and they get bitten every time we add a function or even an optional parameter to an existing function.

So I would like to see the `external`s removed from the .mlis. Bytecode speed is not an issue nowadays.

Comment (xclerc) [reply stdlib/bytes.mli]: I guess @gasche was referring to something like:

```
module F (X : sig
    external f : unit -> unit = "%ignore"
end) = struct
end

module OK = F (struct
    external f : unit -> unit = "%ignore"
  end)

module KO = F (struct
    let f () = ()
  end)
```

**Comment by pmetzger** [comment]
Two comments:

1. Why not include 8 bit quantities as well? Yes, the function to do so is trivial, but it will make some protocol handling code look more uniform.
2. About endianness, I'm always reminded of Rob Pike's insight that code that knows about endianness (other than compilers) is generally badly written. Here's his blog post on the topic. The code is all written in C but it's simple enough not to matter. https://commandcenter.blogspot.com/2012/04/byte-order-fallacy.html

**Comment by dbuenzli** [comment]
> About endianness, I'm always reminded of Rob Pike's insight that code that knows about endianness (other than compilers) is generally badly written.

I don't understand what this is supposed to bring to the discussion. Would you care to expand maybe ?

**Comment by alainfrisch** [comment]
- Are we ok with 21 new functions just in Bytes (12 get, 9 set)?  Should we organize them in submodules (per endianness) as suggested by @nojb?  (I generally prefer having explicit signatures.)
- Assuming there is no runtime overhead, would it be good to pass the endianness choice (ne/le/be) as an argument, in order to divide the number of functions by 3?
- What about adding int8/uint8 variants?
- I believe that using Sys.int_size instead of Sys.word_size (as done in ocplib-endian) is better for JS backends, but that would need to be checked.

**Comment by jhwoodyatt** [comment]
For comparison purposes, I should point out that I have implemented logic for this purpose in my Orsetto project. Start [here](https://bitbucket.org/jhw/orsetto/src/unstable/src/cf/cf_endian.mli) for details if you're interested.

Accordingly, I have no strong opinion on how this pull request might be evaluated. My unsolicited opinions on the various questions discussed above summarized:
- unsafe interfaces are useful in the case where a higher-level interface is ensuring that data format invariants are respected.
- byte-order should not by inferred from the system architecture when formatting data for interchange,  i.e. it should always be exposed in the programming interface without a default.
- formatting IEEE-754 floating point for interchange should provide support for binary16, binary32 and binary64 formats.

**Comment by alainfrisch** [comment]
- I've tried with
````ocaml
val Buffer.add_int64: [`LE|`BE|`NE] -> t -> int64 -> unit

let[@ocaml.inline] add_int64 e b x =
  let new_position = b.position + 8 in
  if new_position > b.length then resize b 8;
  unsafe_set_int64 b.buffer b.position
    (match e, Sys.big_endian with
     | `LE, true | `BE, false -> swap64 x
     | _ -> x);
  b.position <- new_position
````
and thanks to the inline annotation, this gives the same performance than ad hoc functions when endianness is known statically.

- @jhwoodyatt : do you mean that we should not expose "native endianness" functions, and let the user rely on Sys.big_endian to choose between little/big endian variants if he really wants to use the native endianness (e.g. to interact with C datastructures)?  This indeed makes it more explicit which parts of the code depends on the local native endianness.

**Comment by dbuenzli** [comment]
> and thanks to the inline annotation, this gives the same performance than ad hoc functions when endianness is known statically.

I think having all the function is in general more predictable. Except for guaranteed statically known checks done at the toplevel (like `Sys.big_endian`) I'm a little bit wary on relying too much on the compiler optimizations to get the baseline  performance. Suppose I want to make something generic on endianness, with the current proposal I will abstract over the functions. With your last proposal I might abstract over [LE|NE|BE] which (I guess) might give poor performance if I'm not careful.

Personally I don't have a problem with all the additions (`uint8` and `int8` is something is also something I have in that unpublished code if you want to add them) and I personally don't see a huge benefit of gathering those in submodules. Having everything flat, including native endianness, is nice to write and obvious to read in my opinion.

**Comment by jhwoodyatt** [comment]
@alainfrisch no, I mean that you should provide three variations and require the choice made explicitly in the programming interface:

- Little-endian byte order
- Big-endian byte order (or so-called "network byte order")
- System architecture byte order

That's what I did, and I think it was the right choice.

**Comment by jhwoodyatt** [comment]
@alainfrisch no, I mean that you should provide three variations and require the choice made explicitly in the programming interface:

- Little-endian byte order
- Big-endian byte order (or so-called "network byte order")
- System architecture byte order

That's what I did, and I think it was the right choice. Another thing I did was [basically] this:

```
type t = [ `BE | `LE | `SE ]
val create: [< t ] -> (module Endian)
```

Usually, the byte order is known statically. I think if dynamic byte order is really on the menu, then it's best to use a function that selects the right module.

**Comment by dbuenzli** [comment]
@alainfrisch any progress on that one ? I would like to rely on this for further stdlib improvements.

**Comment by alainfrisch** [comment]
Remaining points:
  - [ ] Write the doc for all functions.  **I'd like to get feedback on the existing text before copying them furiously to all other functions**
  - [ ] Port functions from Bytes to String (once the exact design/doc for Bytes is stable)
  - [ ] Add tests

File: stdlib/buffer.mli [hunk 2]
```diff
@@ -177,3 +177,34 @@ val add_seq : t -> char Seq.t -> unit
 val of_seq : char Seq.t -> t
 (** Create a buffer from the generator
     @since 4.07 *)
+
+(** {6 Binary storage of numbers} *)
+
+val add_int16_ne : t -> int -> unit
```
Comment (dbuenzli) [inline stdlib/buffer.mli:231]: I think this should be:
```
(** [add_int16_ne b i] appends a binary native-endian signed 16-bit integer [i] to [b]. @since NEXT_RELEASE *)
```
So the pattern is:
```
(** [add_intXX_EE b i] appends a binary EE-endian signed XX-bit integer [i] to [b]. @since NEXT_RELEASE *)
```

File: stdlib/buffer.mli [hunk 3]
```diff
@@ -177,3 +177,34 @@ val add_seq : t -> char Seq.t -> unit
 val of_seq : char Seq.t -> t
```
Comment (dbuenzli) [inline stdlib/buffer.mli:177]: Please also add `add_uint8`, `add_int8` and `add_uint16_{ne,le,be}`.

File: stdlib/buffer.mli (hunk 2 above)
Comment (mshinwell) [reply stdlib/buffer.mli:231]: I think "EE-endian" would be clearer (and probably more grammatically correct, since it is being used as an adjective).

File: stdlib/bytes.mli [hunk 4]
```diff
@@ -463,6 +463,106 @@ val of_seq : char Seq.t -> t
 (** Create a string from the generator
     @since 4.07 *)
 
+(** {6 Binary storage of integers} *)
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: I would add a comment here that all these functions raise `Invalid_argument` if there's not enough space. Something like:
```
All following function raise [Invalid_argument] if the space needed at index [i] to decode 
or encode the integer is not available.
```

Comment (dbuenzli) [inline stdlib/bytes.mli]: I would add a comment here that all these functions raise `Invalid_argument` if there's not enough space. Something like:
```
All following functions raise [Invalid_argument] if the space needed at index [i] to decode 
or encode the integer is not available.
```

File: stdlib/buffer.mli (hunk 3 above)
Comment (alainfrisch) [reply stdlib/buffer.mli:177]: I don't think we need add_uint*, since all add/set functions consider only the lsb (and so behave the same on signed and unsigned).

File: stdlib/bytes.mli [hunk 5]
```diff
@@ -463,6 +463,106 @@ val of_seq : char Seq.t -> t
...
+
```
Comment (dbuenzli) [inline stdlib/bytes.mli:494]: Please also add `{get,set}_uint8`, `{get,set}_int8`.

File: stdlib/bytes.mli [hunk 6]
```diff
@@ -463,6 +463,106 @@ val of_seq : char Seq.t -> t
...
+val get_uint16_ne : bytes -> int -> int
+(** [get_uint16_ne s n] reads a 16-bit unsigned integer from the byte
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: I think this should be:
```
[get_uint16_ne b i] is [b]'s native-endian unsigned 16-bit integer starting at byte index [i]. @since NEXT_RELEASE
```
so the pattern is: 
```
[get_SSintXX_EE b i] is [b]'s EE-endian SS XX-bit integer starting at byte index [i]. @since NEXT_RELEASE
```

File: stdlib/bytes.mli [hunk 7]
```diff
@@ -463,6 +463,106 @@ val of_seq : char Seq.t -> t
...
+    sequence [s] starting at index [n], in binary form, using the
+    machine native endianness. The result is an integer between 0 and
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: I don't think it's worth spelling out the limits of `int16` and/or `uint16` it should be pretty clear to a programmer what to expect I think.

File: stdlib/bytes.mli [hunk 8]
```diff
@@ -463,6 +463,106 @@ val of_seq : char Seq.t -> t
...
+    native endianness.
+
+    Raise [Invalid_argument] if [n] or [n+7] is not a valid index in
+    [s].
+
+    @since XXX
+*)
+
+
+val set_int16_ne : bytes -> int -> int -> unit
```
Comment (dbuenzli) [inline stdlib/bytes.mli:607]: I think this should be:
```
[set_uint16_ne b i v] sets [b]'s native-endian unsigned 16-bit integer starting at byte index [i] to [v].
```
so the pattern is: 
```
[set_SSintXX_EE b i v] sets [b]'s EE-endian SS XX-bit integer starting at byte index [i] to [v].
```

**Comment by dbuenzli** [comment]
> Write the doc for all functions. I'd like to get feedback on the existing text before copying them furiously to all other functions

Reviewed, see my comments. I think we should keep the doc strings simple using the patterns I mentioned.

File: stdlib/buffer.mli (hunk 2 above)
Comment (alainfrisch) [reply stdlib/buffer.mli:231]: - Don't you want to mention the behavior when [i] does not fit in 16-bit?

- I don't think that mentioning signedness for add/set functions make sense (the functions only truncates the binary representation); signed vs unsigned makes sense when reading, to control whether the result is sign-extended or not.

- Is native-endian a common terminology?  Shouldn't the documentation stresses that the behavior is machine-dependent?

**Comment by alainfrisch** [comment]
@dbuenzli   See my comments above (https://github.com/ocaml/ocaml/pull/1864#discussion_r206776099 in particular).  Pending questions

  - I don't think unsigned variants for add/set functions are useful, except perhaps for API symmetry (they would be aliases to signed variants).  
  - Add/set functions should specify their behavior (truncation).  Do you agree?  Or do you think we should leave the overflow behavior unspecified/undocumented?  Or even check+fail?

**Comment by dbuenzli** [comment]
> * I don't think unsigned variants for add/set functions are useful, except perhaps for API symmetry (they would be aliases to signed variants).

You are absolutely right. But I still think it has the opportunity of making code reading easier. Suppose I read somewhere an unsigned int and then add it to a buffer. The reader maintains in her head that this is an unsigned integer and suddenly she see it added to a buffer with `add_int16` rather than `add_uint16`. This might trigger baseless suspicion about the code if she doesn't remember that only reads matter as far as signedess is concerned. Having `add_uint16` wouldn't have done so.

Since adding those is not a maintenance burden and has the opportunity of clarifying code, I would really be in favour of having them.

> * Add/set functions should specify their behavior (truncation).  Do you agree?  

Yes. Not sure about which behaviour is the best though.

**Comment by alainfrisch** [comment]
> Yes. Not sure about which behaviour is the best though.

Taking the least significant bits is what the current primitives do I believe.  This is what the current version of the PR documents (but not your proposal for a simplified wording, hence my question).  Would the other behavior be to raise?  I think I prefer the current semantics as it leads to more efficient code (and those primitives might specifically be used in low-level code that needs to be fast).

**Comment by dbuenzli** [comment]
> I think I prefer the current semantics as it leads to more efficient code (and those primitives might specifically be used in low-level code that needs to be fast).

Yes. In any case I wouldn't find it a surprising behaviour. 

> (but not your proposal for a simplified wording, hence my question).

Note that my suggested doc string patterns only need to be clarified for 8-bit and 16-bit. So I'd use the following for `32` and `64`:

```ocaml
(** [add_intXX_EE b i] appends to [b] the binary EE-endian signed XX-bit 
    integer [i]. @since NEXT_RELEASE *)
```

And for `8` and `16`:

```ocaml
(** [add_intXX_EE b i] appends to [b] the binary EE-endian signed XX-bit 
    integer [i] truncated to XX-bits. @since NEXT_RELEASE *)
```

**Comment by alainfrisch** [comment]
> [add_intXX_EE b i] appends to [b] the binary EE-endian signed XX-bit  integer [i] truncated to XX-bits

It sounds weird to truncate to XX-bits an XX-bit integer.  What about:

> [add_int16_EE b i] appends to [b] the 16 least significant bits of the signed integer [i], using EE-endian encoding.

(for int8, we don't need to specify endianness.)

**Comment by dbuenzli** [comment]
> It sounds weird to truncate to XX-bits an XX-bit integer.

Maybe the following would be more clear then:
```
[add_intXX_EE b i] appends to [b] a binary EE-endian signed XX-bit 
integer resulting from truncating [i] to XX-bits. @since NEXT_RELEASE 
```
I prefer to mention that we are appending an integer rather than bits, since that's what the function name says.

**Comment by alainfrisch** [comment]
@dbuenzli I've pushed a commit using the original style you suggested for docstrings in Bytes, mentioning how "extra bits" in `int` compared to 8- and 16-bit storage are handled in the section header (truncate + zero/sign extension).  Does that work for you?  If so, I'll apply the same style for Buffer.

File: stdlib/bytes.mli [hunk 9]
```diff
@@ -463,11 +463,37 @@ val of_seq : char Seq.t -> t
...
+
+    All following functions raise [Invalid_argument] if the space
     needed at index [i] to decode or encode the integer is not
-    available. *)
+    available.
+
+    Little-endian (resp. big-endian) encoding means that least
+    (resp. most) significant bytes are stored first.  Big-endian is
+    also known as network byte order.  Native-endian encoding is
+    either little-endian or big-endian depending on the platform.
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: I would link the "depending" to `Sys.big_endian`. That is `{{!Sys.big_endian}depending}`.

**Review by dbuenzli — COMMENTED** [review]
Looks good to me.

**Comment by alainfrisch** [comment]
I've dealt with the Buffer module as well (doc + aliases for uint8/uint16), added a test, and adapted BytesLabels.

I think this is now ready for a full review.

(I will add a Changes entry afterwards to avoid repeated conflicts; and also replace NEXT_RELEASE at the same time.)

One might consider adding get functions to String as well, but I prefer to do that in another PR.

File: stdlib/buffer.ml [hunk 10]
```diff
@@ -299,3 +299,61 @@ let of_seq i =
...
+external unsafe_set_int32 : bytes -> int -> int32 -> unit = "%caml_bytes_set32u"
+external unsafe_set_int64 : bytes -> int -> int64 -> unit = "%caml_bytes_set64u"
+external swap16 : int -> int = "%bswap16"
+external swap32 : int32 -> int32 = "%bswap_int32"
+external swap64 : int64 -> int64 = "%bswap_int64"
+
+
+let add_int8 b x =
+  let pos = b.position in
+  if pos >= b.length then resize b 1;
```
Comment (dbuenzli) [inline stdlib/buffer.ml]: Would be nice to use the same style as the following functions.

File: stdlib/buffer.ml [hunk 11]
```diff
@@ -299,3 +299,61 @@ let of_seq i =
...
+let add_int16_be b x =
+  add_int16_ne b (if Sys.big_endian then x else swap16 x)
+
+let add_int32_le b x =
+  add_int32_ne b (if Sys.big_endian then swap32 x else x)
+
+let add_int32_be b x =
+  add_int32_ne b (if Sys.big_endian then x else swap32 x)
+
+
```
Comment (dbuenzli) [inline stdlib/buffer.ml]: Unecessary blank line.

File: stdlib/buffer.mli [hunk 12]
```diff
@@ -177,3 +177,108 @@ val add_seq : t -> char Seq.t -> unit
...
 (** Create a buffer from the generator
     @since 4.07 *)
+
+(** {6 Binary encoding of integers} *)
```
Comment (dbuenzli) [inline stdlib/buffer.mli]: Maybe this should be done as a separate PR but all these level heading 6 in this file should be turned into 1.

File: stdlib/buffer.mli [hunk 13]
```diff
@@ -177,3 +177,108 @@ val add_seq : t -> char Seq.t -> unit
...
+
+(** The functions in this section encode integers stored
+    in binary form into a buffer.
```
Comment (dbuenzli) [inline stdlib/buffer.mli]: This sentence is a bit misleading. I would at least remove the "stored in", otherwise it seems the integers are in binary form and it's unclear how you encode them. We don't really care about how integers are stored, the type tells you.

Changing a bit more the sentence I would suggest:  "The functions in this section append binary encodings of integers to buffers".

File: stdlib/buffer.mli [hunk 14]
```diff
@@ -177,3 +177,108 @@ val add_seq : t -> char Seq.t -> unit
...
+    either little-endian or big-endian depending on {!Sys.big_endian}.
+
+    32-bit and 64-bit integers are directly represented using
+    [int32] and [int64] types, which can be interpreted either
+    as signed or unsigned numbers.
+
+    8-bit and 16-bit integers are represented using the [int] type,
+    which has more bits than the binary encoding.  Functions that
+    encode these values truncate their inputs to their least
+    significant bytes.  Signed and unsigned variants are synonyms of
```
Comment (dbuenzli) [inline stdlib/buffer.mli]: I think the last sentence can be dropped.

File: stdlib/buffer.mli [hunk 15]
```diff
@@ -177,3 +177,108 @@ val add_seq : t -> char Seq.t -> unit
...
+    each other.
+    *)
+
+val add_uint8 : t -> int -> unit
+(** [add_uint8 b i] appends a binary unsigned XX-bit integer [i] to
```
Comment (dbuenzli) [inline stdlib/buffer.mli]: XX -> 8

File: stdlib/buffer.mli [hunk 16]
```diff
@@ -177,3 +177,108 @@ val add_seq : t -> char Seq.t -> unit
...
+    [b].
+    @since NEXT_RELEASE
+*)
+
+val add_int8 : t -> int -> unit
+(** [add_int8 b i] appends a binary signed XX-bit integer [i] to
```
Comment (dbuenzli) [inline stdlib/buffer.mli]: XX -> 8

File: stdlib/bytes.ml [hunk 17]
```diff
@@ -368,3 +368,86 @@ let of_seq i =
...
+let get_int64_le b i =
+  if Sys.big_endian then swap64 (get_int64_ne b i)
+  else get_int64_ne b i
+
+let get_int64_be b i =
+  if not Sys.big_endian then swap64 (get_int64_ne b i)
+  else get_int64_ne b i
+
+let set_int8 b i x =
+  set b i (Char.unsafe_chr x)
```
Comment (dbuenzli) [inline stdlib/bytes.ml]: I'm unsure about the `unsafe_chr` bit. I would have `x land 0xFF`.

File: stdlib/bytes.mli [hunk 18]
```diff
@@ -463,6 +463,206 @@ val of_seq : char Seq.t -> t
 (** Create a string from the generator
     @since 4.07 *)
 
+(** {6 Binary encoding/decoding of integers} *)
+
+(** The functions in this section decode or encode integers stored
+    in binary form in byte sequences.
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: Here again I would change the sentence. Maybe "The functions in this section binary encode and decode integers to and from byte sequences".

File: stdlib/bytes.mli [hunk 19]
```diff
@@ -463,6 +463,206 @@ val of_seq : char Seq.t -> t
...
+    All following functions raise [Invalid_argument] if the space
+    needed at index [i] to decode or encode the integer is not
+    available.
+
+    Little-endian (resp. big-endian) encoding means that least
+    (resp. most) significant bytes are stored first.  Big-endian is
+    also known as network byte order.  Native-endian encoding is
+    either little-endian or big-endian depending on {!Sys.big_endian}.
+
+    32-bit and 64-bit integers are directly represented using
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: I would drop the "directly" and "represented by the" rather "represented using".

File: stdlib/buffer.mli [hunk 20]
```diff
@@ -177,3 +177,108 @@ val add_seq : t -> char Seq.t -> unit
...
+
+(** The functions in this section encode integers stored
+    in binary form into a buffer.
+
+    Little-endian (resp. big-endian) encoding means that least
+    (resp. most) significant bytes are stored first.  Big-endian is
+    also known as network byte order.  Native-endian encoding is
+    either little-endian or big-endian depending on {!Sys.big_endian}.
+
+    32-bit and 64-bit integers are directly represented using
```
Comment (dbuenzli) [inline stdlib/buffer.mli]: I would drop the "directly" and "represented by the" rather "represented using".

File: stdlib/bytes.mli [hunk 21]
```diff
@@ -463,6 +463,206 @@ val of_seq : char Seq.t -> t
...
+    [int32] and [int64] types, which can be interpreted either
+    as signed or unsigned numbers.
+
+    8-bit and 16-bit integers are represented using the [int] type,
+    which has more bits than the binary encoding.  These extra bits
+    are handled as follows:
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: The two next paragraphs would benefit being in a list: 
```
{ul 
{- ... }}
{- ... }}
```

File: stdlib/bytes.mli [hunk 22]
```diff
@@ -463,6 +463,206 @@ val of_seq : char Seq.t -> t
...
+
+    Functions that decode signed (resp. unsigned) 8-bit or 16-bit
+    integers represented as [int] values sign-extend
+    (resp. zero-extend) their result.
+
+    Functions that encode 8-bit or 16-bit integers represented as
+    [int] values truncate their input to their least significant
+    bytes.  Signed and unsigned variants are synonyms of each other
```
Comment (dbuenzli) [inline stdlib/bytes.mli]: Again I would drop the last sentence.

File: stdlib/bytesLabels.mli [hunk 23]
```diff
@@ -308,6 +308,205 @@ val of_seq : char Seq.t -> t
 (** Create a string from the generator
     @since 4.07 *)
 
+(** {6 Binary encoding/decoding of integers} *)
```
Comment (dbuenzli) [inline stdlib/bytesLabels.mli]: I guess this is a c&p so I won't bother going through it. The comments above apply.

File: testsuite/tests/lib-bytes/binary.ml [hunk 24]
```diff
@@ -0,0 +1,230 @@
...
+  lazy (Bytes.set_int64_le b 3 0L) |> err;
+  lazy (Bytes.set_int64_ne b 3 0L) |> err;
+  lazy (Bytes.set_int64_be b 3 0L) |> err;
+  lazy (Bytes.get_int64_le b 3) |> err;
+  lazy (Bytes.get_int64_ne b 3) |> err;
+  lazy (Bytes.get_int64_be b 3) |> err;
+  ()
+
+
+let () =
```
Comment (dbuenzli) [inline testsuite/tests/lib-bytes/binary.ml]: These tests rather belong to `testsuite/tests/lib-buffer/test.ml`

**Review by dbuenzli — COMMENTED** [review]
I left a few documentation nits, one question and a suggestion to move the buffer tests to the test file for buffers. Otherwise seems good to me.

**Comment by gasche** [comment]
(I'm prepared to approve the PR, on the behalf of @dbuenzli (without reviewing myself), once you are both satisfied.)

File: stdlib/buffer.ml (hunk 10 above)
Comment (alainfrisch) [reply stdlib/buffer.ml]: Ok.

File: stdlib/buffer.mli (hunk 12 above)
Comment (alainfrisch) [reply stdlib/buffer.mli]: I'm changing from 6 to 1, but only for the additions related to this PR.

File: stdlib/buffer.mli (hunk 14 above)
Comment (alainfrisch) [reply stdlib/buffer.mli]: OK.

File: stdlib/bytes.ml (hunk 17 above)
Comment (alainfrisch) [reply stdlib/bytes.ml]: Do you mean `Char.unsafe_chr (x land 0xff)`?   I think the current code is ok, since unsafe_chr is `%identity` and the primitive truncates to the least significant byte.  The extra `land 0xFF` is probably not very costly; do you think I should add it?   Alternatively, I could expose directly
````
external set_int8 : bytes -> int -> int -> unit = "%bytes_safe_set"
````
(i.e. the same external as for Bytes.set, but with an `int` argument instead of `char`) to be coherent with other primitives.  Let me know what you prefer.

File: stdlib/bytes.mli (hunk 18 above)
Comment (alainfrisch) [reply stdlib/bytes.mli]: Ok.

**Comment by alainfrisch** [comment]
Just one minor pending point about the use of Char.unsafe_chr.  Waiting for @dbuenzli feedback on that one.

File: stdlib/bytes.ml (hunk 17 above)
Comment (dbuenzli) [reply stdlib/bytes.ml]: > Do you mean `Char.unsafe_chr (x land 0xff)`? I think the current code is ok, since unsafe_chr is `%identity` and the primitive truncates to the least significant byte.

Yes that's what I meant. I guess you are right but OTOH you are clearly violating the expectations of `Char.unsafe_chr` and injecting values in the type that should not exist there.

Maybe I'm a bit too pedantic since this injection is only transient but simply using `%bytes_safe_set` makes things cleaner without being abusive to `Char.t`.

Comment (alainfrisch) [reply stdlib/bytes.ml]: Ok, see latest commit.  I'm now using directly %-primitives for int8 versions, which is more regular.

**Review by dbuenzli — APPROVED** [review]
I'd say we are good to go. Thanks for putting the work in this.

**Comment by alainfrisch** [comment]
Thanks for your help with the design and documentation!

**Review by gasche — APPROVED** [review]
